### PR TITLE
fix `Override the href property of the SVGImageElement.prototype` (close #1502)

### LINKED ISSUE #1502: Override the href property of the SVGImageElement.prototype

### PR DESCRIPTION
**Pull request by Farfurix**
Changes:

1. Override ```SVGImageElement.prototype.href```(get) property that returns [```SVGAnimatedString```](https://developer.mozilla.org/ru/docs/Web/API/SVGAnimatedString) object, add context element property to returned object
2. Override ```SVGAnimatedString.animVal```(get), ```SVGAnimatedString.baseVal```(get, set) properties

Links:

https://developer.mozilla.org/en-US/docs/Web/API/SVGImageElement:
> The ```SVGImageElement``` interface corresponds to the ```<image>``` element.

https://developer.mozilla.org/en-US/docs/Web/SVG/Element/image:
> The HTML spec defines <image> as a synonym for  while parsing HTML. This specific element and its behavior only apply inside SVG documents or inline SVG.

https://developer.mozilla.org/en-US/docs/Web/SVG/Tutorial/SVG_In_HTML_Introduction:
> To add a linked image with DOM methods to an embedded SVG element, one has to use ```setAttributeNS``` to set ```href```. Like in the following example:
> ```
> var img = document.createElementNS("http://www.w3.org/2000/svg", "image");
> img.setAttributeNS("http://www.w3.org/1999/xlink", "xlink:href", "move.png");
> ```

https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/xlink:href:
> An ```<IRI>``` reference to the image source. On the ```<image>``` element, the ```xlink:href``` attribute is animatable.

https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/xlink:href:
> SVG 2 removed the need for the xlink namespace, so instead of ```xlink:href``` you should use ```href```

### CODE REVIEW

**Comment by Farfurix** [comment]
@testcafe-build-bot \retest

**Comment by Farfurix** [comment]
@testcafe-build-bot \retest

**Comment by Farfurix** [comment]
@testcafe-build-bot \retest

File: src/client/sandbox/node/window.js [hunk 1]
```diff
@@ -51,6 +52,8 @@ const ALLOWED_SERVICE_WORKER_PROTOCOLS  = ['https:', 'wss:', 'file:'];
 const ALLOWED_SERVICE_WORKER_HOST_NAMES = ['localhost', '127.0.0.1'];
 const JAVASCRIPT_MIME_TYPES             = ['text/javascript', 'application/javascript', 'application/x-javascript'];
 
+const CONTEXT_IMAGE_ELEMENT = 'hammerhead|contextImageElement';
```
Comment (LavrovArtem) [inline src/client/sandbox/node/window.js]: `'hammerhead|context-image-element'`

File: src/client/sandbox/node/window.js [hunk 2]
```diff
@@ -724,6 +727,56 @@ export default class WindowSandbox extends SandboxBase {
         this._overrideUrlPropDescriptor('protocol', nativeMethods.anchorProtocolGetter, nativeMethods.anchorProtocolSetter);
         this._overrideUrlPropDescriptor('search', nativeMethods.anchorSearchGetter, nativeMethods.anchorSearchSetter);
 
+        overrideDescriptor(window.SVGImageElement.prototype, 'href', {
+            getter: function () {
+                const imageHref = nativeMethods.imageHrefGetter.call(this);
+
+                return Object.defineProperty(imageHref, CONTEXT_IMAGE_ELEMENT, { value: this });
```
Comment (LavrovArtem) [inline src/client/sandbox/node/window.js]: I think we need to add `this[CONTEXT_IMAGE_ELEMENT]` condition. Also add `configurable: true` to descriptor.
And use the `defineProperty` function from `nativeMethods`.

File: test/client/fixtures/sandbox/code-instrumentation/getters-test.js [hunk 3]
```diff
@@ -188,6 +188,29 @@ test('HTMLElement.style', function () {
...
+test('image (GH-1502)', function () {
+    var imageNS       = document.createElementNS('http://www.w3.org/2000/svg', 'image');
+    var nativeImageNS = nativeMethods.createElementNS.call(document, 'http://www.w3.org/2000/svg', 'image');
+    var url           = 'https://google.com:1888/index.html?value#yo';
+    var proxyUrl      = urlUtils.getProxyUrl(url);
+
+    imageNS.href.baseVal = url;
+    nativeMethods.baseValSetter.call(nativeMethods.imageHrefGetter.call(nativeImageNS), url);
+
+    strictEqual(nativeMethods.getAttribute.call(imageNS, 'href-hammerhead-stored-value'), url);
```
Comment (LavrovArtem) [inline test/client/fixtures/sandbox/code-instrumentation/getters-test.js]: Add to top
`var DomProcessor      = hammerhead.get('../processing/dom');`
and use
`DomProcessor.getStoredAttrName('href')`

File: test/client/fixtures/sandbox/code-instrumentation/getters-test.js [hunk 4]
```diff
@@ -188,6 +188,29 @@ test('HTMLElement.style', function () {
...
+    strictEqual(nativeMethods.animValGetter.call(imageNS.href), proxyUrl);
+    strictEqual(nativeMethods.baseValGetter.call(imageNS.href), proxyUrl);
+    strictEqual(imageNS.href.animVal, urlUtils.parseProxyUrl(proxyUrl).destUrl);
```
Comment (LavrovArtem) [inline test/client/fixtures/sandbox/code-instrumentation/getters-test.js]: ` strictEqual(imageNS.href.animVal, nativeImageHref.animVal);`

File: test/client/fixtures/sandbox/code-instrumentation/getters-test.js [hunk 5]
```diff
@@ -188,6 +188,29 @@ test('HTMLElement.style', function () {
...
+    strictEqual(imageNS.href.baseVal, urlUtils.parseProxyUrl(proxyUrl).destUrl);
+    strictEqual(imageNS.href['hammerhead|contextImageElement'], imageNS);
```
Comment (LavrovArtem) [inline test/client/fixtures/sandbox/code-instrumentation/getters-test.js]: I think it's unnecessary.

File: test/client/fixtures/sandbox/code-instrumentation/getters-test.js [hunk 6]
```diff
@@ -188,6 +189,24 @@ test('HTMLElement.style', function () {
...
+    svgImageNS.href.baseVal = url;
+    nativeMethods.svgAnimStrBaseValSetter.call(nativeMethods.svgImageHrefGetter.call(nativeSvgImageNS), url);
+
+    var hrefStroredAttrName = DomProcessor.getStoredAttrName.call(svgImageNS, 'href');
+
+    strictEqual(nativeMethods.getAttribute.call(svgImageNS, hrefStroredAttrName), url);
+    strictEqual(nativeMethods.svgAnimStrAnimValGetter.call(svgImageNS.href), proxyUrl);
+    strictEqual(nativeMethods.svgAnimStrBaseValGetter.call(svgImageNS.href), proxyUrl);
+    strictEqual(svgImageNS.href.animVal, nativeSvgImageNS.href.animVal);
+    strictEqual(svgImageNS.href.baseVal, nativeSvgImageNS.href.baseVal);
```
Comment (LavrovArtem) [inline test/client/fixtures/sandbox/code-instrumentation/getters-test.js]: add
`strictEqual(svgImageNS.getAttributeNS(ns, 'href'), nativeMethods.getAttributeNS.call(nativeSvgImageNS, ns, 'href');`

File: test/client/fixtures/sandbox/code-instrumentation/setters-test.js [hunk 7]
```diff
@@ -134,6 +134,17 @@ test('anchor', function () {
         strictEqual(anchorWithNotSupportedProtocol.origin, nativeMethods.anchorOriginGetter.call(etalonAnchorWithNotSupportedProtocol));
 });
 
+test('image (GH-1502)', function () {
```
Comment (LavrovArtem) [inline test/client/fixtures/sandbox/code-instrumentation/setters-test.js]: I don't understand what this test do.

Comment (Farfurix) [reply test/client/fixtures/sandbox/code-instrumentation/setters-test.js]: It's partial duplication of getter test, so i delete it.

**Comment by Farfurix** [comment]
@testcafe-build-bot \retest

File: test/server/data/page/src.html [hunk 8]
```diff
@@ -104,6 +104,12 @@
     <use xlink:href="http://domain.com/test.svg#rect"></use>
 </svg>
 <div xlink:href="http://domain.com/test.svg#rect"></div>
+<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink= "http://www.w3.org/1999/xlink">
+    <image xlink:href="https://mdn.mozillademos.org/files/6457/mdn_logo_only_color.png">
```
Comment (miherlosev) [inline test/server/data/page/src.html]: Do we need real images in test data?
I guess we can use fake urls for this case.

**Pull request merged**